### ISSUE TITLE
`bst artifact checkout --tar` should allow compression

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#1048]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/1048)
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Jun 21, 2019, 16:08

## Background

[//]: # (Provide a background or the root/source that justifies this task or action.)

`bst artifact checkout --tar` is provided as a simple way to produce a tarball output of a BuildStream element - this means that we can retrieve our output in a distributable form without having to create a dedicated element for that. However, if we want to also compress our artifact (as is typical for this sort of thing), we need to run our compression tool separately. This is annoying (although admittedly a very minor thing).

Python allows compressing tarfiles, so this is fairly trivial to achieve. There is a question of reproducibility here, since the details of the compression algorithm will depend on the python version, but given that we already allow python to create a tarfile (whose precise format can change slightly) in the first place this seems to be acceptable.

## Task description

[//]: # (Short summary of the action to be executed)

We should make BuildStream compress tarfiles when the user wants to do so. I suggest making BuildStream "do what I mean" by default, in other words, if I create a file named "hello.tar.gz", I expect it to be compressed with gzip. We'd mimick GNU-tar's behavior this way, making the `--tar` option a bit more intuitive at the same time. We should probably also add compression flags to allow users to override BuildStream's detection.

In a nutshell, we should:

* [ ] Make BuildStream compress artifacts checked out using `bst artifact checkout --tar` according to the file name by default.
* [ ] Add flags to allow users to override the default to the command (stay consistent with GNU tar flags).

*hint*: The relevant code mostly lives in `__frontend/cli.py` and `_stream.py` (line 515 and onwards).

## Acceptance Criteria

[//]: # (Acceptance criteria should follow the S.M.A.R.T. principle https://en.wikipedia.org/wiki/SMART_criteria )

* [ ] BuildStream provides an interface that allows checking out artifacts as a compressed tarball.

----
[//]: # (To review information about possible relevant labels for this issue please view the list of labels: https://gitlab.com/BuildStream/buildstream/labels)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @beckyella16]](https://gitlab.com/beckyella16) on Jun 21, 2019, 16:48

I can have a look at this.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Jun 25, 2019, 12:29

This would be a nice addition!

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Jun 25, 2019, 12:43

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Jul 2, 2019, 12:40

I wonder if the simplest approach from a CLI point of view would to be make --tar take two args, the first staying as Path() filename which we'd default to selecting a compression type from the extension if given, and the second optional positional being click.Choice for the compression type with no change to the default behaviour if not given as to not break existing calls? The API for click seems to support [Tuples as Multi Value Options](https://click.palletsprojects.com/en/7.x/options/#tuples-as-multi-value-options) but I'm not a click expert so it might not fit our usecase here.

This would stop the naive approach of having a '--compression' (or similar) flag to exist that requires `--tar` to have been given, and handling that combination not being given correctly.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jennis]](https://gitlab.com/jennis) on Jul 3, 2019, 10:57

mentioned in merge request !1413

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @beckyella16]](https://gitlab.com/beckyella16) on Jul 4, 2019, 11:41

closed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @beckyella16]](https://gitlab.com/beckyella16) on Jul 4, 2019, 11:45

reopened

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @beckyella16]](https://gitlab.com/beckyella16) on Jul 4, 2019, 12:13

mentioned in merge request !1451

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tmewett]](https://gitlab.com/tmewett) on Aug 9, 2019, 10:28

closed
